provider/oci.go

```diff
@@ -0,0 +1,298 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"io/ioutil"
+	"strings"
+
+	"github.com/oracle/oci-go-sdk/common"
+	"github.com/oracle/oci-go-sdk/dns"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/kubernetes-incubator/external-dns/endpoint"
+	"github.com/kubernetes-incubator/external-dns/plan"
+)
+
+const ociRecordTTL = 300
+
+// OCIAuthConfig holds connection parameters for the OCI API.
+type OCIAuthConfig struct {
+	Region      string `yaml:"region"`
+	TenancyID   string `yaml:"tenancy"`
+	UserID      string `yaml:"user"`
+	PrivateKey  string `yaml:"key"`
+	Fingerprint string `yaml:"fingerprint"`
+	Passphrase  string `yaml:"passphrase"`
+}
+
+// OCIConfig holds the configuration for the OCI Provider.
+type OCIConfig struct {
+	Auth          OCIAuthConfig `yaml:"auth"`
+	CompartmentID string        `yaml:"compartment"`
+}
+
+// OCIProvider is an implementation of Provider for Oracle Cloud Infrastructure
+// (OCI) DNS.
+type OCIProvider struct {
+	client ociDNSClient
+	cfg    OCIConfig
+
+	domainFilter DomainFilter
+	zoneIDFilter ZoneIDFilter
+	dryRun       bool
+}
+
+// ociDNSClient is the subset of the OCI DNS API required by the OCI Provider.
+type ociDNSClient interface {
+	ListZones(ctx context.Context, request dns.ListZonesRequest) (response dns.ListZonesResponse, err error)
+	GetZoneRecords(ctx context.Context, request dns.GetZoneRecordsRequest) (response dns.GetZoneRecordsResponse, err error)
+	PatchZoneRecords(ctx context.Context, request dns.PatchZoneRecordsRequest) (response dns.PatchZoneRecordsResponse, err error)
+}
+
+// NewOCIProvider initialises a new OCI DNS based Provider.
+func NewOCIProvider(configFile string, domainFilter DomainFilter, zoneIDFilter ZoneIDFilter, dryRun bool) (*OCIProvider, error) {
+	contents, err := ioutil.ReadFile(configFile)
```

Comment:
Entirely subjective (and potentially overkill) but could be factored out to make it testable i.e. 

config, err := loadConfig(configFile) 

where loadConfig(configFile string) (*OCIConfig, error)